ADR: Enable github discussions to enable broader community (#19479)
quarkusio/quarkus: adr/0001-community-discussions.adoc

= Setup Quarkus Github Discussions to enable the broader Community

* Status: proposed
* Date: 2021-06-26

== Context and Problem Statement

Quarkus community is growing and until now we've catered very well for core contributors and initial early consumers. 

We have multiple communication channels: https://github.com/quarkusio/quarkus[issues], https://groups.google.com/g/quarkus-dev?pli=1[quarkus-dev mailing list], https://quarkusio.zulipchat.com[zulip], https://stackoverflow.com/questions/tagged/quarkus[stackoverflow]

> Isues are great for bugs/feature work. Mailing list for design conversations between developers, chat for watercooler style discussions and stackoverflow for user questions.

Reviewer: ```suggestion
Issues are great for bugs/feature work. Mailing list for design conversations between developers, chat for watercooler-style discussions and stackoverflow for user questions.
```

This setup has issues though, some are:

> - zulip chat is used for a lot of users questions but none of that is easily searchable/discoverable so it is very synchronous,

Reviewer: ```suggestion
- Zulip chat is used for a lot of user's questions but none of that is easily searchable/discoverable so it is very synchronous,
```

- people reported that they don't feel okey posting on quarkus-dev or zulips as it is seems focused on dev work and not so much about community events, jobs, conferences, etc.

> - its hard to monitor as contributor who wants to help answer/ask questions.

Reviewer: ```suggestion
- it's hard to monitor as contributor who wants to help answer/ask questions.
```

> How can we improve this situaton and enable the broader community to more easily ask questions and find answers - without it all be relying on just a few Quarkus core contributors?

Reviewer 2: I think this misses a bit of positioning regarding SO.

== Scenarios (optional)

> User wants to locate a answer to a question - zulip chats does not show up in google search; stackoverflow might but might not have an answer.

Reviewer 2: That's what we try to avoid and we constantly monitor SO. Not great as it's a manual process, but we do our best. 

SO gets well indexed by google and now, if you asked the right question, Google displays the answer directly.

Author: interesting factoid - githubs wikis are blocked in github.com/robots.txt but discussions are not.

About SO - its not going away but its a chore to monitor and many feel its getting more and more unfriendly because of all its requirements.

I think we'll keep having SO and zulip users but I honestly think github discussions is a better fit.

Author: found this thread about SEO on GitHub discussions: https://github.com/github/feedback/discussions/3318

summary:
make sure quarkus.io and other pages link to the discussions page.
give it 90 days to pick up properly by google
success!

until then search from GitHub UI is better.

> Contributor want to arrange or attend an event around Quarkus - where do he post/look for info on that ?

Reviewer: ```suggestion
Contributors want to arrange or attend an event around Quarkus - where do they post/look for info on that ?
```

> 

Reviewer 2: That's a good point - and it's more like a forum kind of thing. Do you have an example of GH Discussions that is used that way? In Mutiny, we use it as a SO clone.

Author: yes - I see github discussions big win for us to be more a forum with optional enablement of Q/A sections.

The "biggest" one is https://github.com/vercel/next.js/discussions with its show and tell 

even the default categories made by github has non q/a sections.

Contributor/committer are trying to follow new questions and issues - what does he need to track to do that ? today it is alot and it is all different notification sytems and mainly its a big "firehose" with no good ability to filter out what you don't need.

> User looking for a job involving Quarkus - where do he go look ?

Reviewer: ```suggestion
User looking for a job involving Quarkus - where do they go look ?
```

////
 as well as a lot of social platform presence https://twitter.com/quarkusio/[twitter], https://www.facebook.com/quarkusio/[facebook], https://www.linkedin.com/groups/13789086/[linkedin], https://www.youtube.com/quarkusio[youtube] and https://www.linkedin.com/groups/13789086/[reddit].
////

== Considered options

* stackoverflow will not go away - thus for this current proposal it does not include Q/A section.

> * "I can't track more places" is a very valid worry; the advantage of discussions is that it shows up in your normal github notifcations which you can mute, filter,etc. like anything else on Github. Thus it gives you more control rather than just adding more noise.

Reviewer 3: ```suggestion
* "I can't track more places" is a very valid worry; the advantage of discussions is that it shows up in your normal github notifications which you can mute, filter,etc. like anything else on Github. Thus it gives you more control rather than just adding more noise.
```

> * "I can't track more places" is a very valid worry; the advantage of discussions is that it shows up in your normal github notifcations which you can mute, filter,etc. like anything else on Github. Thus it gives you more control rather than just adding more noise.

Reviewer 3: My main worry about this is that people will disable the notifications and Georgios and I will have to ping the appropriate person. We don't want that. Thus why we need the triaging system from the beginning IF we open a Quarkus Community section - even if you say you don't want usage question, I'm pretty sure they will end up there.

Reviewer 2: As discussed, if we can integrate it into the triage system that's a BIG plus.

Author: so good news - we can update labels:

I just used the following snippet to add a discussion label in jbangdev discussion forum:

```
mutation {
  addLabelsToLabelable(input: {labelableId: "D_kwDODbWxfs4AN-Qb", labelIds: ["MDU6TGFiZWwyMjM4OTQ3MDM2"]}) {
    clientMutationId
  }
}
```

turns out there is a generic API to update anything "labelable":  https://docs.github.com/en/graphql/reference/mutations#addlabelstolabelable

figured out with help from the GitHub discussion PM https://twitter.com/tweetsbyevi/status/1456366292823261190

so this is no longer a blocker - yay!

> * "I can't track more places" is a very valid worry; the advantage of discussions is that it shows up in your normal github notifications which you can mute, filter,etc. like anything else on Github. Thus it gives you more control rather than just adding more noise.

Reviewer: ```suggestion
* "I can't track more places" is a very valid worry; the advantage of discussions is that it shows up in your normal GitHub notifications which you can mute, filter,etc. like anything else on GitHub. Thus it gives you more control rather than just adding more noise.
```

=== quarkusio-community mailing list
Not proposed since it is just another waterhose of info with no good ability to ignore/consume as needed/wanted.

=== enable discussions on github.com/quarkusio/quarkus-insights

> Considered as quarkus insights is an existing "community" and let us have separate permissons for moderators/triage but some suggested it might be hard to find.

Reviewer 2: ```suggestion
Considered as quarkus insights is an existing "community" and let us have separate permissions for moderators/triage but some suggested it might be hard to find.
```

=== enable discussions on github.com/quarkusio/community
Downside is that labels etc. would need to be kept in sync and that searches in github.com/quarkusio would not show up community/discussions.
It could show up if searched on github.com/quarkus level though.
Could be considered as allow us to have separate permissions for moderator/triagers separate from commit access - need to check if https://docs.github.com/en/organizations/managing-access-to-your-organizations-repositories/repository-permission-levels-for-an-organization[github docs] on permissions levels work for us.

== Decision

> Enable GitHub discussions feature on `quarkusio/quarkusio` with the following initial categories:

Reviewer 3: ```suggestion
Enable GitHub discussions feature on `quarkusio/quarkus` with the following initial categories:
```

> - Announcements (post only by admins)

Reviewer 3: I won't post release announcements in several channels so either they will go there or to quarkus-dev if you open both (+ Twitter). That being said, I don't think it's worth posting release announcements here.

So just be aware: if you open this, don't ask me in two months to also post the release announcements there :).

Reviewer 2: I think we automated that in Mutiny: https://github.com/smallrye/smallrye-mutiny/discussions/categories/announcements. But I don't know if @jponge does it or if it's automatic.

Author: github releases can automatically be posted there and be linked to have conversations on the release.

Reminds me of another usecase  i've seen applied - use github discussion as comment source for the blog: https://artsy.github.io/blog/2017/07/15/Comments-are-on/

Not something for day one but having all this in one "bucket" gives some nice simplifications.

> - Quarkus Comunity (post by anyone, general discussions)

Reviewer 3: ```suggestion
- Quarkus Community (post by anyone, general discussions)
```

> - Quarkus Community (post by anyone, general discussions)

Reviewer 4: Community is vast, wat would be the content ?
- links to article, video, ...
- usefull tips
- questions
- seeks for help
- ???

Author: anything that does not fit into the other categories baiscally.

Author: maybe just rename it to General ?

- Quarkus Insights Episodes (show notes and offline comments)

> - Quarkus World Tour and other Events (Setup and announce of interest or other events)

Reviewer 4: ```suggestion
- Quarkus Events (Setup and announce of interest or other events)
```
The World Tour is just one events, if World Tour is going to last forever maybe it deserve it's own category

Author: agreed.

> - Jobs and Opportunities (post by anyone)

Reviewer 3: One thing I would like would be a place where people can present themselves if they wish. E.g. one thread per person with the name of the person and they describe what they are doing with Quarkus.

I'm not sure people would want to do that but when I interact with people I often ask myself about the context or what they're doing and things like that.

Again no idea if anyone would participate to that.

Author: +1; some will. What should we call it ? or maybe just a post with certain label in quarkus community ?

Reviewer: +1, sounds like a good opportunity to know who is using Quarkus in their jobs

Reviewer 5: Did it go away? I don't see "Introductions" in the list?

Reviewer 4: +1 to introduction and yes it's not on the list

Author: git push mishab - its there now.

== Consequences

With this the conversation(s) in Quarkus community is no longer only about development on quarkus-dev mailing list nor  just by Quarkus core contributors. That should enable broader participation and let Quarkus community grow beyond just "the code".

> We don't (yet) add a Q/A section to this as Zulip Users and Stackoverflow still exist.

Reviewer: ```suggestion
We don't (yet) add a Q/A section to this as Zulip users and Stack Overflow still exist.
```

> We don't (yet) add a Q/A section to this as Zulip Users and Stackoverflow still exist.

Reviewer 4: The first sections talks a lot about user Q&A, the fact they sometime don't feel compfortable to post on Zulip and the fac that Zulip is not searchable so I don't understand this decision.
Maybe a "Seek for help" section would be needed.

Author: was missing a commit - updated.

> Downsides: "Another notification sinkhole"; yes - it is another additional location to keep an eye on but since this is hosted at Github you can use the unified notifcation mechanism and system; especially you can just subscribe or unsubrcibe based on  what is best for the invidiual community member.

Reviewer 3: ```suggestion
Downsides: "Another notification sinkhole"; yes - it is another additional location to keep an eye on but since this is hosted at Github you can use the unified notification mechanism and system; especially you can just subscribe or unsubscribe based on what is best for the individual community member.
```

> 

Reviewer 3: There's a good chance Quarkus Community will get the Q/A. You might not want to put it in this document but for me, the condition to opening a Quarkus Community section is a proper automatic triaging.

Author: The more I think about it I think we should make a section for Quarkus Usage and Quarkus Extensions with the Q&A mode enabled. i'll update the pr with that and position it in relation to SO.

> Downsides: "Another notification sinkhole"; yes - it is another additional location to keep an eye on but since this is hosted at Github you can use the unified notification mechanism and system; especially you can just subscribe or unsubscribe based on what is best for the individual community member.

Reviewer: ```suggestion
Downsides: "Another notification sinkhole"; yes - it is another additional location to keep an eye on but since this is hosted at GitHub you can use the unified notification mechanism and system; especially you can just subscribe or unsubscribe based on what is best for the individual community member.
```
